rasa_sdk/forms.py

```diff
@@ -726,57 +726,90 @@ async def extract_custom_slots(
         """
         custom_slots = []
-        slots_to_extract = await self.required_slots(dispatcher, tracker, domain)
-
-        if slots_to_extract is None:
-            return []
+        slots_to_extract = await self.required_slots(
+            self.slots_mapped_in_domain(domain), dispatcher, tracker, domain
+        )
 
         for slot_name in slots_to_extract:
-            method_name = f"extract_{slot_name.replace('-', '_')}"
-            extract_method = getattr(self, method_name, None)
+            custom_slots += await self._extract_slot(
+                slot_name, dispatcher, tracker, domain
+            )
+        return custom_slots
 
-            if not extract_method:
-                warnings.warn(
-                    f"No method '{method_name}' found for slot "
-                    f"'{slot_name}'. Skipping extraction for this slot."
-                )
-                continue
+    async def _extract_slot(
+        self,
+        slot_name: Text,
+        dispatcher: "CollectingDispatcher",
+        tracker: "Tracker",
+        domain: "DomainDict",
+    ) -> List[EventType]:
+        method_name = f"extract_{slot_name.replace('-', '_')}"
+        also_mapped_in_domain = slot_name in self.slots_mapped_in_domain(domain)
+        extract_method = getattr(self, method_name, None)
+
+        if not extract_method and not also_mapped_in_domain:
+            warnings.warn(
+                f"No method '{method_name}' found for slot "
+                f"'{slot_name}'. Skipping extraction for this slot."
+            )
+            return []
+
+        if not extract_method:
+            return []
 
-            extracted = await utils.call_potential_coroutine(
-                extract_method(dispatcher, tracker, domain)
+        if also_mapped_in_domain:
+            warnings.warn(
+                f"Slot '{slot_name}' is mapped in the domain and your custom "
+                f"action defines '{method_name}'. It is recommended to define a "
+                f"slot mapping in either of both places."
             )
 
-            if extracted:
-                custom_slots += [
-                    SlotSet(slot, value) for slot, value in extracted.items()
-                ]
-            else:
-                warnings.warn(
-                    f"Cannot extract `{slot_name}`: make sure the extract method "
-                    f"returns the correct output."
-                )
+        extracted = await utils.call_potential_coroutine(
+            extract_method(dispatcher, tracker, domain)
+        )
 
-        return custom_slots
+        if extracted:
+            return [SlotSet(slot, value) for slot, value in extracted.items()]
+
+        warnings.warn(
+            f"Cannot extract `{slot_name}`: make sure the extract method "
+            f"returns the correct output."
+        )
+        return []
 
     async def required_slots(
         self,
+        slots_mapped_in_domain: List[Text],
```

Comment:
@wochinge what do you think about renaming this to `slots_with_predefined_mappings` or similar? I'm not 100% certain and would like to get someone else's opinion, but it's true that in the docs I've been trying to remove the mentions of slots "mapped in the domain".

Reply:
I think in the documentation it made sense in combination with the link. I think in this case `domain` makes it clearer for now.